pkg/backends/calico/client.go

```diff
@@ -811,6 +885,258 @@ func (c *client) onUpdates(updates []api.Update) {
+							communityValues = append(communityValues, definedCommunity.Value)
+							break
+						}
+					}
+				} else {
+					communityValues = append(communityValues, c)
+				}
+			}
+			if cidrVersion == 4 {
+				ipv4PrefixToAdvertise = append(ipv4PrefixToAdvertise, bgpPrefix{
+					CIDR:        cidr,
+					Communities: communityValues,
+				})
+			} else if cidrVersion == 6 {
+				ipv6PrefixToAdvertise = append(ipv6PrefixToAdvertise, bgpPrefix{
+					CIDR:        cidr,
+					Communities: communityValues,
+				})
+			}
+		}
+
+		ipv4Communities, ok := json.Marshal(ipv4PrefixToAdvertise)
+		if ok != nil {
+			log.Warningf("Error while marshalling BGP communities. %#v", ok)
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKeyIPv4, string(ipv4Communities)))
+
+		ipv6Communities, ok := json.Marshal(ipv6PrefixToAdvertise)
+		if ok != nil {
+			log.Warningf("Error while marshalling BGP communities. %#v", ok)
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKeyIPv6, string(ipv6Communities)))
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKeyIPv4))
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKeyIPv6))
+	}
+}
+
+func (c *client) getListenPortKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKey := getBGPConfigKey("listen_port", key)
+
+	if v3res != nil && v3res.Spec.ListenPort != 0 {
+		switch key.(type) {
+		case model.NodeBGPConfigKey:
+			c.nodeListenPorts[getNodeName(v3res.Name)] = int(v3res.Spec.ListenPort)
+		case model.GlobalBGPConfigKey:
+			c.globalListenPort = int(v3res.Spec.ListenPort)
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, strconv.Itoa(int(v3res.Spec.ListenPort))))
+	} else {
+		switch k := key.(type) {
+		case model.NodeBGPConfigKey:
+			delete(c.nodeListenPorts, getNodeName(k.Nodename))
+		case model.GlobalBGPConfigKey:
+			c.globalListenPort = 0
+		}
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKey))
+	}
+}
+
+func (c *client) getASNumberKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKey := getBGPConfigKey("as_num", key)
+
+	if v3res != nil && v3res.Spec.ASNumber != nil {
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, v3res.Spec.ASNumber.String()))
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKey))
+	}
+}
+
+func (c *client) getServiceExternalIPsKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKey := getBGPConfigKey("svc_external_ips", key)
+
+	if v3res != nil && v3res.Spec.ServiceExternalIPs != nil {
+		// We wrap each Service external IP in a ServiceExternalIPBlock struct to
+		// achieve the desired API structure, unpack that.
+		ipBlocks := v3res.Spec.ServiceExternalIPs
+		if len(ipBlocks) != 0 {
+			ipCidrs := make([]string, len(ipBlocks))
+			for i, ipBlock := range ipBlocks {
+				ipCidrs[i] = ipBlock.CIDR
+			}
+			c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, strings.Join(ipCidrs, ",")))
+		} else {
```

Comment:
Need a comment here.  Why are we handling `len(v3res.Spec.ServiceExternalIPs) == 0` differently from `v3res.Spec.ServiceExternalIPs == nil` ?

Reply:
BGPConfigUpdateProcessor in libcalico-go used to send updates with `nil` value if `len(v3res.Spec.ServiceExternalIPs) == 0`, I tried replicating that here. But both those conditions you pointed have same result (aka existing cache entry gets removed), I'll update it .